lib/import_export/base_exporter.rb

```diff
@@ -0,0 +1,161 @@
+
+      self
+    end
+
+    # export groups that are mentioned in category permissions
+    def export_category_groups
+      groups = []
+      group_names = []
+      auto_group_names = Group::AUTO_GROUPS.keys.map(&:to_s)
+
+      @export_data[:categories].each do |c|
+        c[:permissions_params].each do |group_name, _|
+          group_names << group_name unless auto_group_names.include?(group_name.to_s)
+        end
+      end
+
+      group_names.uniq!
+      return [] if group_names.empty?
+
+      Group.where(name: group_names).find_each do |group|
+        attrs = GROUP_ATTRS.inject({}) { |h, a| h[a] = group.send(a); h }
+        attrs[:user_ids] = group.users.pluck(:id)
+        groups << attrs
+      end
+
+      groups
+    end
+
+    def export_category_groups!
+      @export_data[:groups] = export_category_groups
+
+      self
+    end
+
+    def export_group_users
+      user_ids = []
+
+      @export_data[:groups].each do |g|
+        user_ids += g[:user_ids]
+      end
+
+      user_ids.uniq!
+      return [] if user_ids.empty?
+
+      users = User.where(id: user_ids)
+      export_users(users)
+    end
+
+    def export_group_users!
+      @export_data[:users] = export_group_users
+
+      self
+    end
+
+    def export_topics
+      data = []
+
+      topics.each do |topic|
+        puts topic.title
+
+        topic_data = TOPIC_ATTRS.inject({}) { |h, a| h[a] = topic.send(a); h; }
+        topic_data[:posts] = []
+
+        topic.ordered_posts.find_each do |post|
+          h = POST_ATTRS.inject({}) { |h, a| h[a] = post.send(a); h; }
+          h[:raw] = h[:raw].gsub('src="/uploads', "src=\"#{Discourse.base_url_no_prefix}/uploads")
+          topic_data[:posts] << h
+        end
+
+        data << topic_data
+      end
+
+      data
+    end
+
+    def export_topics!
+      @export_data[:topics] = export_topics
+
+      self
+    end
+
+    def export_topic_users
+      topic_ids = @export_data[:topics].pluck(:id)
+
+      users = TopicUser.includes(user: [:user_profile]).where(topic_id: topic_ids).joined_includes_values.pluck(:user)
+      users.uniq!
+
+      export_users(users)
+    end
+
+    def export_topic_users!
+      @export_data[:users] = export_topic_users
+
+      self
+    end
+
+    def export_users(users)
+      data = []
+
+      users.each do |u|
+        x = USER_ATTRS.inject({}) { |h, a| h[a] = u.send(a); h; }
+        x.merge(bio_raw: u.user_profile.bio_raw,
+                website: u.user_profile.website,
+                location: u.user_profile.location)
+        data << x
+      end
+
+      data
+    end
+
+    def default_filename_prefix
+      raise "Overwrite me!"
+    end
+
+    def save_to_file(filename = nil)
+      output_basename = filename || File.join("#{default_filename_prefix}-#{Time.now.strftime("%Y-%m-%d-%H%M%S")}.json")
+      File.open(output_basename, "w:UTF-8") do |f|
+        f.write(@export_data.to_json)
+      end
+      puts "Export saved to #{output_basename}"
```

Comment:
Is this required? We should avoid adding `puts` statement into our code base. You can log stuff within the rake task itself instead.

Reply:
This one is really important I guess. We are printing exported file's name in console window so the user can copy it.